src/test/java/hudson/plugins/copyartifact/CopyArtifactWorkflowTest.java

```diff
@@ -0,0 +1,115 @@
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package hudson.plugins.copyartifact;
+
+import hudson.model.Run;
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.junit.Assert;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.JenkinsRule;
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * @author <a href="mailto:tom.fennelly@gmail.com">tom.fennelly@gmail.com</a>
+ */
+public class CopyArtifactWorkflowTest {
+
+    @Rule
+    public JenkinsRule jenkinsRule = new JenkinsRule();
+
+    @Test
+    public void test_simpleUntriggeredCopy() throws Exception {
+        // create "project_1" with an archived artifact...
+        WorkflowJob project_1 = createWorkflow("project_1",
+                "sh 'echo hello > hello.txt'; " +
+                "step([$class: 'ArtifactArchiver', artifacts: 'hello.txt', fingerprint: true])");
+        WorkflowRun b = jenkinsRule.assertBuildStatusSuccess(project_1.scheduleBuild2(0));
+        assertArtifactInArchive(b);
+
+        // Now lets try copy the artifact from "project_1" to "project_2"
+        WorkflowJob project_2 = createWorkflow("project_2",
+                "step([$class: 'CopyArtifact', projectName: 'project_1', filter: 'hello.txt']); " +
+                "step([$class: 'ArtifactArchiver', artifacts: 'hello.txt', fingerprint: true]);");
+        b = jenkinsRule.assertBuildStatusSuccess(project_2.scheduleBuild2(0));
+        assertArtifactInArchive(b);
+    }
+
+    @Test
+    public void test_upstreamCopyFromDownstream() throws Exception {
+        // create "project_1" with an archived artifact...
+        createWorkflow("project_1",
+                "sh 'echo hello > hello.txt'; " +
+                "step([$class: 'ArtifactArchiver', artifacts: 'hello.txt'])");
+
+        // Create "project_2", which triggers build of "project_1"
+        WorkflowJob project_2 = createWorkflow("project_2",
+                "build ('project_1'); " +
+                "step ([$class: 'CopyArtifact', projectName: 'project_1', filter: 'hello.txt']); " +
+                "step ([$class: 'ArtifactArchiver', artifacts: 'hello.txt', fingerprint: true]);");
+
+        // run project_2 and check that was able to copy from project_1
+        WorkflowRun b = jenkinsRule.assertBuildStatusSuccess(project_2.scheduleBuild2(0));
+        assertArtifactInArchive(b);
+    }
+
+    @Test
+    public void test_downstreamCopyFromUpstream() throws Exception {
+        // create "project_1" with an archived artifact...
+        WorkflowJob project_1 = createWorkflow("project_1",
+                "sh 'echo hello > hello.txt'; " +
+                "step([$class: 'ArtifactArchiver', artifacts: 'hello.txt']); " +
+                "build ('project_2'); ");
```

Comment:
Testing that a workflow can successfully copy an artifact from a build that triggered it is not relevant?